File: scripts/build_static_data.mjs
/**
 * Build static data for Netlify: HTC index from tolltariffen JSONs, split large files by chapter.
 * Run from repo root: node scripts/build_static_data.mjs
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, '..');
const DATA_RAW = path.join(REPO_ROOT, 'data', 'raw');
const DATA = path.join(REPO_ROOT, 'data');
const OUT = path.join(REPO_ROOT, 'frontend', 'data');

function walkCommodities(nodes) {
  if (!Array.isArray(nodes)) nodes = nodes ? [nodes] : [];
  const out = [];
  for (const n of nodes) {
    if (!n || typeof n !== 'object') continue;
    if (n.type === 'commodity') {
      const code = String(n.id || n.hsNumber || '').trim();
      const name = String(n.item || '').trim();
      if (code) out.push({ code, name, description: '' });
    }
    for (const key of ['sections', 'chapters', 'divisions', 'headings', 'subchapters']) {
      if (!(key in n)) continue;
      const child = n[key];
      if (Array.isArray(child)) out.push(...walkCommodities(child));
      else if (child != null) out.push(...walkCommodities([child]));
    }
  }
  return out;
}

function buildHtcIndex() {
  const p = path.join(DATA_RAW, 'customstariffstructure.json');
  if (!fs.existsSync(p)) {
    console.log('Missing data/raw/customstariffstructure.json');
    return;
  }
  const data = JSON.parse(fs.readFileSync(p, 'utf8'));
  const commodities = walkCommodities(data.sections || []);
  const seen = new Set();
  const index = [];
  for (const { code, name } of commodities) {
    if (seen.has(code)) continue;
    seen.add(code);
    index.push({ code, name, description: '' });
  }
  if (!fs.existsSync(OUT)) fs.mkdirSync(OUT, { recursive: true });
  fs.writeFileSync(path.join(OUT, 'htc_index.json'), JSON.stringify(index), 'utf8');
  console.log('Wrote htc_index.json with', index.length, 'HTCs.');
}

function splitByChapter(filePath, outSubdir) {
  if (!fs.existsSync(filePath)) {
    console.log('Missing', filePath);
    return;
  }
  const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  const obj = typeof data === 'object' && data !== null && !Array.isArray(data) ? data : {};
  const byChapter = {};
  for (const [code, value] of Object.entries(obj)) {
    if (code.length >= 2) {
      const ch = code.slice(0, 2);
      if (!byChapter[ch]) byChapter[ch] = {};
      byChapter[ch][code] = value;
    }
  }
  const outDir = path.join(OUT, outSubdir);
  if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });
  for (const [ch, chunk] of Object.entries(byChapter).sort()) {
    fs.writeFileSync(path.join(outDir, ch + '.json'), JSON.stringify(chunk), 'utf8');
  }
  console.log('Wrote', outSubdir, 'with', Object.keys(byChapter).length, 'chapters.');
}

function copyJson(name) {
  const src = path.join(DATA, name);
  if (!fs.existsSync(src)) {
    console.log('Missing', src);
    return;
  }
  if (!fs.existsSync(OUT)) fs.mkdirSync(OUT, { recursive: true });
  fs.copyFileSync(src, path.join(OUT, name));
  console.log('Copied', name);
}

function main() {
  if (!fs.existsSync(OUT)) fs.mkdirSync(OUT, { recursive: true });
  buildHtcIndex();
  splitByChapter(path.join(DATA, 'best_zero_countries.json'), 'best_zero');
  splitByChapter(path.join(DATA, 'ratetradeagreements_index.json'), 'ratetradeagreements');
  copyJson('country_names.json');
  copyJson('landgroups_map.json');
  console.log('Static data build done.');
}

main();

File: scripts/build_static_data.py
"""
Build static data for Netlify: HTC search index from tolltariffen JSONs,
split large files by chapter for on-demand load. Run from repo root.
"""
from __future__ import annotations
import json
import shutil
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
DATA_RAW = REPO_ROOT / "data" / "raw"
DATA = REPO_ROOT / "data"
OUT = REPO_ROOT / "frontend" / "data"


def walk_commodities(nodes) -> list[tuple[str, str]]:
    if not isinstance(nodes, list):
        nodes = [nodes] if nodes else []
    out = []
    for n in nodes:
        if not isinstance(n, dict):
            continue
        if n.get("type") == "commodity":
            code = str(n.get("id") or n.get("hsNumber") or "").strip()
            name = (n.get("item") or "").strip()
            if code:
                out.append((code, name))
        for key in ("sections", "chapters", "divisions", "headings", "subchapters"):
            if key not in n:
                continue
            child = n[key]
            if isinstance(child, list):
                out.extend(walk_commodities(child))
            elif child is not None:
                out.extend(walk_commodities([child]))
    return out


def build_htc_index():
    path = DATA_RAW / "customstariffstructure.json"
    if not path.exists():
        print("Missing data/raw/customstariffstructure.json")
        return
    data = json.loads(path.read_text(encoding="utf-8"))
    commodities = walk_commodities(data.get("sections", []))
    seen = set()
    index = []
    for code, name in commodities:
        if code in seen:
            continue
        seen.add(code)
        index.append({"code": code, "name": name, "description": ""})
    OUT.mkdir(parents=True, exist_ok=True)
    (OUT / "htc_index.json").write_text(json.dumps(index, ensure_ascii=False, indent=0), encoding="utf-8")
    print(f"Wrote htc_index.json with {len(index)} HTCs.")


def split_by_chapter(file_path: Path, out_subdir: str, key_in_file: str | None = None):
    """Read a JSON object keyed by HTC code; write one file per chapter (01..99)."""
    if not file_path.exists():
        print(f"Missing {file_path}")
        return
    data = json.loads(file_path.read_text(encoding="utf-8"))
    if key_in_file:
        data = data.get(key_in_file, data)
    if not isinstance(data, dict):
        return
    by_chapter: dict[str, dict] = {}
    for code, value in data.items():
        if len(code) >= 2:
            ch = code[:2]
            by_chapter.setdefault(ch, {})[code] = value
    out_dir = OUT / out_subdir
    out_dir.mkdir(parents=True, exist_ok=True)
    for ch, chunk in sorted(by_chapter.items()):
        (out_dir / f"{ch}.json").write_text(json.dumps(chunk, ensure_ascii=False), encoding="utf-8")
    print(f"Wrote {out_subdir}/ with {len(by_chapter)} chapters.")


def copy_json(name: str):
    src = DATA / name
    if not src.exists():
        print(f"Missing {src}")
        return
    OUT.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, OUT / name)
    print(f"Copied {name}")


def main():
    OUT.mkdir(parents=True, exist_ok=True)
    build_htc_index()
    split_by_chapter(DATA / "best_zero_countries.json", "best_zero")
    split_by_chapter(DATA / "ratetradeagreements_index.json", "ratetradeagreements")
    copy_json("country_names.json")
    copy_json("landgroups_map.json")
    print("Static data build done.")


if __name__ == "__main__":
    main()
